webapi/src/main/scala/org/knora/webapi/util/date/CalendarDateUtilV2.scala

```diff
@@ -0,0 +1,543 @@
+      * @return a [[CalendarDateV2]] representing the date.
+      */
+    def parse(dateStr: String, calendarName: CalendarNameV2): CalendarDateV2 = {
+        // Get the era, if provided.
+
+        val dateStringSplitByEra: Array[String] = dateStr.split(StringFormatter.EraSeparator)
+
+        val maybeEra: Option[DateEraV2] = dateStringSplitByEra.length match {
+            case 1 =>
+                // No era given. In the Gregorian or Julian calendar, the era defaults to CE.
+                calendarName match {
+                    case _: CalendarNameGregorianOrJulian => Some(DateEraCE)
+                    case _ => None
+                }
+
+            case 2 =>
+                // An era was given. Parse it.
+                val eraStr: String = dateStringSplitByEra(1)
+                Some(DateEraV2.parse(eraStr, throw BadRequestException(s"Invalid date era: $eraStr")))
+
+            case _ => throw BadRequestException(s"Invalid date: $dateStr")
+        }
+
+        // Get the year, month, and day.
+
+        val dateSegments = dateStringSplitByEra(0).split(StringFormatter.PrecisionSeparator)
+
+        try {
+            dateSegments.length match {
+                case 1 => // year precision
+                    CalendarDateV2(
+                        calendarName = calendarName,
+                        year = dateSegments.head.toInt,
+                        maybeMonth = None,
+                        maybeDay = None,
+                        maybeEra = maybeEra
+                    )
+
+                case 2 => // month precision
+                    CalendarDateV2(
+                        calendarName = calendarName,
+                        year = dateSegments.head.toInt,
+                        maybeMonth = Some(dateSegments(1).toInt),
+                        maybeDay = None,
+                        maybeEra = maybeEra
+                    )
+
+                case 3 => // day precision
+                    CalendarDateV2(
+                        calendarName = calendarName,
+                        year = dateSegments.head.toInt,
+                        maybeMonth = Some(dateSegments(1).toInt),
+                        maybeDay = Some(dateSegments(2).toInt),
+                        maybeEra = maybeEra
+                    )
+
+                case _ => throw BadRequestException(s"Invalid date: $dateStr")
+            }
+        } catch {
+            case e: IllegalArgumentException => throw BadRequestException(s"Invalid date '$dateStr': $e")
+        }
+    }
+}
+
+/**
+  * Represents a date range consisting of two instances of [[CalendarDateV2]]. Both instances must have the same
+  * calendar.
+  *
+  * @param startCalendarDate the start of the range.
+  * @param endCalendarDate the end of the range.
+  */
+case class CalendarDateRangeV2(startCalendarDate: CalendarDateV2, endCalendarDate: CalendarDateV2) {
+    if (startCalendarDate.calendarName != endCalendarDate.calendarName) {
```

Comment:
You could also check if `startCalendarDate ` is before or equals `endCalendarDate`.

Reply:
Yes, good point.

Reply:
Done in https://github.com/dhlab-basel/Knora/pull/1006/commits/c1977f8e8caa8bb301921baf988d1eedd6857a2e.

Reply:
great, thx.